lib/src/main/java/com/knuddels/jtokkit/TokenEncoder.java

```diff
@@ -1,80 +1,196 @@
 package com.knuddels.jtokkit;
 
+
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
-import java.util.Optional;
-import java.util.function.Function;
-
-/**
- * A TokenEncoder is used to encode and decode tokens. It is initialized with a map
- * that contains the decoded tokens as keys and the encoded tokens as values. The
- * TokenEncoder can then be used to encode and decode tokens.
- */
-class TokenEncoder {
-
-    private final Map<ImmutableByteArray, Integer> decodedToEncoded = new HashMap<>();
-
-    /**
-     * Creates a new TokenEncoder with the given input map. The keys of the map are
-     * the decoded tokens and the values are the encoded tokens.
-     *
-     * @param input the input map
-     */
-    public TokenEncoder(Map<ImmutableByteArray, Integer> input) {
-        this(input, Function.identity());
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+final class TokenEncoder {
+    public static final int DUMMY_RANK = Integer.MAX_VALUE;
+    public static final int MAX_RANK = Integer.MAX_VALUE - 1;
+    private final Map<ImmutableByteArray, Integer>[] encoders;
```

Comment:
The encoders map is split by input byte array size so that we're only querying small maps